man/man6/lsys.6
.TH LSYS 6 "Version 1, Release 4" "W Window System" "W PROGRAMS"
.SH NAME
lsys \- L-system parser for W
.SH SYNOPSIS
.nf
lsys [<options>] < rules
.fi
.SH DESCRIPTION
.I Lsys
is an L-system parser.  L-systems are grammar based fractals.  The
grammars consist of certain tokens that denote graphical actions.  Some
of the tokens can then be recursively expanded (to a specified depth)
following the rule(s) in the grammar.  The images (after the resulting
new grammar has been translated to graphics) are self-similar at
different magnifications and therefore fractals.
.SH ARGUMENTS
.TP
\fILsys\fP options are:
\fB-w\fP \fIwindow width\fP		(image is auto-scaled
.br
\fB-h\fP \fIwindow height\fP		 to this size)
.br
\fB-d\fP \fIrecursion depth\fP		(0-4 are reasonable)
.br
\fB-a\fP \fI360 / turn angle\fP		(3->, in degrees)
.PP
The rules file may include the following things in the order given:
.IP
\(bu angles (360 / turn angle to use)
.br
\(bu axiom  (initial rule)
.br
\(bu recursion rules
.br
\(bu comments
.TP
Recursion rules have the following syntax:
<token> = <rule>.
.PP
Of the tokens used in the rules, following are converted to graphics
instructions:
.IP
.I f
: draw forward
.br
.I F
: move forward
.br
.I +
: turn clockwise		(see the angle argument)
.br
.I -
: turn anti-clockwise
.br
.I [
: store position		(used for branching)
.br
.I ]
: restore position
.PP
White space (spaces, tabs, newlines) in the input will be ignored.
.PP
Lsys stops input parsing at the first non-rule string (second non-space
character not '='), before which you can have as many rules as you wish.
.PP
You could also give the angle and rules 'interactively' (quit input with
carriage return (to flush stdin) and ctrl-D (EOF)) although without
prompt that might not be very comfortable though...
.SH EXAMPLES
The lsys default, classical van Koch's snowflake, is:
.IP
.nf
6
f++f++f
f=f-f++f-f
.fi
.TP
Explanation:
\(bu angle used = 360/6 = 60 degrees
.br
\(bu axiom (initial rule) draws a triangle
.br
\(bu rule changes every line segment to '_/\\_' recursively
.PP
Use different depth values (0-) and see for yourself how the rule
"f-f++f-f" ('_/\\_') is applied to every occurrence of 'f' (straight
line):
.IP
.nf
lsys -d 1 < /dev/null
.fi
.PP
Other grammars to try out are (with depths about 2->):
.TP
Plant:
.nf
12
f
f=f[+f]f[-f]f
.fi
.TP
Bush:
.nf
16
f
f=ff+[+f-f-f]-[-f+f+f]
.fi
.TP
Dragon curve (depths 8->):
.nf
4
x
x=x+yf+
y=-fx-y
.fi
.TP
Peano space fill:
.nf
4
x
x=xfyfx+f+yfxfy-f-xfyfx
y=yfxfy-f-xfyfx+f+yfxfy
.fi
.TP
Square Sierpinsky:
.nf
4
f+f+f+f
f=ff+f+f+f+ff
.fi
.PP
As you see the grammars can include untranslatable tokens.  You can use
them to compose more complex grammar expansion rules.
.SH TECHNICAL NOTES
Because lsystems' fractal dimensions may differs for different grammars
/ recursion depths, I have to calculate the whole grammar to check image
dimensions for auto-scaling.  That may take some time before lsys starts
to draw to the opened window.
.PP
For example, using depth 5 on the last grammar example would expand it
into 67228 (4*7*7*7*7*7) 'f' tokens (= lines, which take 16 bytes each
to transmit to W server).  Also the line length given to 'f' token would
be less than one pixel with the default window size.
.PP
Because of those plant models, I decided that the starting direction
should be up.
.SH LITERATURE
See any general book on fractals...
.PP
Nowadays 3D, raytraced, genetic, mutating, multilevel etc.  enhanced
L-system programs are used to create realistic looking 3D plants, fire
etc.  for film-industry, environment simulations and other hi-tech
visualizations.
.SH AUTHOR
Eero Tamminen, t150315@cc.tut.fi
.SH LAST CHANGES
Feb 1996 by Eero


wt/CHANGES
++oddie 9/2000:
- fixed listbox/filesel resizing & added minimum size setopt.

++bsittler 7/2000:
- Fixed vt color/reverse and attribute resetting
- Add HTTP/1.1 'Host' header to Wetscape requests
- Other minor changes (netgames, wvt, waudio)

++jcarlson 7/2000:
- WTINYSCREEN changes

++eero, 5/1999:
- Fixed key handling for special keys in Wt/getstring.c & wedit/wedit.c.

++eero, 1/1999:
- Added support for new W special keys so that now on W platforms which
implement special key mappings, you don't need to use Emacs shortcuts
with text widgets (or remap your console cursor keys)...


------ notes for version released with W R1PL0 ----------

ADDITIONAL WIDGETS

- Arrow, a simple button with an arrow drawn into it.

- Dial, a dial widget where you can position a needle into desired
  angle.  This isn't in the W toolkit library itself as then you would
  need to link math library to all of your Wt programs.  You'll find it
  in the warzone game tar-ball.

- Edittext, mainly intended for simple one page stuff, but as this
  spaghetti can do some additional stuff like paragraph formatting,
  there's a real abuse built on this, the WEdit.  As edittext widget is
  large and buggy, it's in the WEdit directory instead of the toolkit.

- Menu, a strip of buttons from which popups pop-up.  Both the popups
  and menus use a structure defined in Wt.h instead of strings that Kay
  was using.  The reason is that I want callbacks to follow
  automatically hierarchically opened popups.

- Popup, hierarchical popups.  Can include subpopups, images, lines and
  in addition items can also be disabled and checked.

- Range, a scrollbar composite with which you can select a value between
  given limits.

- Select, a getstring widget coupled with an arrow openable listbox from
  which items can be selected into a getstring widget.


OTHER CHANGES

- vt can now do everything wterm  can do + user can rebind the colors
  in wtrc.

- Added '.wtconfig' file and made makefiles use the W compilation
  configuration file '.config' in parent directory.

- wt_do_event() doesn't call wt_dispatch_event() anymore with
  EVENT_GADGET/GADGET_EXIT event as those might not have valid window
  pointer.  If you want to check for abnormal (unboud) exit, check the
  wt_return value.  GADGET_EXIT returns a negative number and
  GADGET_CLOSE (in shell widget) a positive number now.

- Added xterm mouse events, reverse mode, fd/bgcolor setting + improved
  double/triple click timing in VT widget.

- Aliases a couple new tags into others in HTML widget and added
  HTML header (ignored) and TITLE parsing.

- Widget event processing functions return now unrecognized or
  non-processed WEVENTs or NULL instead of long.  Assumption is that if
  function returns WEVENT, the widget whose action it was, is still
  around.  If in doubt, use NULL.  Unrecognized key events should be
  returned for focus change though.  (actually it could still have been
  an int, but returning a WEVENT offers a possibility for event mapping
  widgets.)

- Added wt_global structure containing Wt configuration variables and
  made widgets to use it.  wt_init() will now read these variables from
  ~/.wtrc + a list of (variable, value) pairs.  Widget specific
  initialization code in their init() functions may read these with the
  wt_variable() function.

- HTML widget fonts can be configured by above method.

- Added wt_loadfont() function which will use the user specified Wt
  default if widget doesn't specify some font parameter.  It also exits
  program if Wt default fonts can't be loaded, because otherwise
  unavailable fonts may chrash naive programs.

- Changed widgets to reflect the new w_loadfont() syntax (NULL/zero
  values tell Wlib/server to use server defaults).

- Added new WT_FONTSIZE define.  The earlier WT_FONT setopt sets now the
  font *family* name, new one sets the font size.

- Getopting WT_FONT returns WFONT* pointer also for above reason.

- Removed 'fontname' members from widget structures.  Needed information
  and more can be gotten from the 'font' member if needed.

- Fileselector has now separate lists for directory and file names.
  Both of course have an autolocator (type & return in addition to point
  & click).

- Wt_keys.h defines values for some special keys like cursors etc.  Half
  of the defines are Emacs style and half what the special keys itself
  give or are in use on other platforms.

- Changed Wt option defines.  There were a couple overlapping ones.


------- Kay's original notes: ----------------------------------------------

- added some missing close() and truncate the file to zero length before
  writing a new selection in selection.c.

- better selection handling in VT widget.

- made everything resizeable.

- applied patches by eero and fixed some more bugs myself.

- implemented a simple clipboard and changed getstring and vt widget to use it.

- fixed several serious bugs in the html widget.

- wrote VT widget.

- added an option to the icon editor to sync edit field with display
  field (simon).

- added missing special chars to html widget (simon).

- added code to the fileselector to set a default file name.

- lots of changes by eero (button.c, fileselect.c, listbox.c, getstring.c), some
  fixes to his changes by me.

- added missing wt_ungetfocus() to listbox and getstring.

- wrote Tk-like packer widget.

- new fileselect widget and testprogram by eero. Added better filename
  matching (eeros version was a joke, really).

- getstring and listbox updates by eero.

- changed the scrollbar widget's user interface to become more user friendly.
  See scrollbar.doc for details on how to use it. Also changed viewport and
  listbox to use the new interface.

- added listbox_focus(). Otherwise listbox cannot receive keyboard focus...

- big bugfix: when wt_delete() is called on a widget all pending timers must
  be canceled! (affected list, iconedit and getstring).


games/lsys.c
/*
 * lsys.c, a part of the W Window System
 *
 * Copyright (C) 1996 by Eero Tamminen
 *
 * This package is free software; you can redistribute it and/or modify it
 * under the terms specified in the docs/COPYRIGHTS file coming with this
 * package.
 *
 * -- a primitive L-system parser and some basic turtle graphics for W
 */

#include <Wlib.h>
#include <math.h>		/* sin(), cos(), M_PI */
#include <stdio.h>
#include <stdlib.h>		/* atoi() */
#include <string.h>		/* mem*() */

/* window defines */
#define WIN_NAME	" W-Lsys "
#define WIN_PROPERTIES	(W_MOVE | W_TITLE | W_CLOSE)
#define DEF_SIZE	256
#define MIN_SIZE	64

/* window ID */
WWIN *Win;

/* parser limit defines */
#define MAX_DEPTH	16	/* max. recursion depth for parsing */
#define MAX_ANGLES	36
#define MAX_RULES	16

typedef struct
{
  double x;
  double y;
} MinMax;

typedef struct
{
  int dir;			/* turtle heading (angle array index) */
  double penx, peny;		/* turtle position */
} TURTLE;


/* global graphics variables */
double Distance;
double Qsin[MAX_ANGLES];
double Qcos[MAX_ANGLES];
int Angles, Scaling;
MinMax Min, Max;		/* for scaling */

int StkIdx;			/* Turtle state (Store) stack index */
TURTLE Store[MAX_DEPTH];	/* grammar -> graphics mapping stack */
TURTLE Turtle;			/* for drawing */


/* prototypes */
void	help(const char *name);
const char *args(int argc, char *argv[], int *depth, int *angle,
		 int *width, int *height);
char	*get_rules(int *angle);
void	parse(const char *axiom, const char *rules, int depth);
void	draw(char token);
const char *win_init(int width, int height);
void	start_drawing(void);
void	check_events(long timeout);


int main(int argc, char *argv[])
{
  int
    idx,
    depth = 3,			/* recursion depth */
    width = DEF_SIZE,		/* window width */
    height= DEF_SIZE;		/* window height */
  const char
    *rules,
    *error;

  Angles = 6;
  if(!(rules = get_rules(&Angles)))
  {
    fprintf(stderr, "%s: invalid rules\n", *argv);
    return -1;
  }
  if((error = args(argc, argv, &depth, &Angles, &width, &height)))
  {
    fprintf(stderr, "%s: %s\n", *argv, error);
    help(*argv);
    return -1;
  }

  if((error = win_init(width, height)))
  {
    fprintf(stderr, "%s: %s\n", *argv, error);
    return -1;
  }

  /* precalculate angles (so that 0 degrees is up) */
  for(idx = 0; idx < Angles; idx++)
  {
    Qsin[idx] = qsin(360.0 / Angles * idx + 90);
    Qcos[idx] = qcos(360.0 / Angles * idx + 90);
  }

  /* setup variables for scaling */
  Max.x = Max.y = Min.x = Min.y = Turtle.penx = Turtle.peny = 0.0;
  Distance = 1.0; Scaling = 1;
  StkIdx = Turtle.dir = 0;

  parse(&rules[1], &rules[rules[0]+1], depth);
  check_events(0L);

  /* Scales Distance / Turtle position values
   * so that the image fits into the window
   */
  if((Max.x - Min.x) > (Max.y - Min.y))
    Distance = (double)(width - width/8) / (Max.x - Min.x);
  else
    Distance = (double)(height - height/8) / (Max.y - Min.y);
  Turtle.penx = (double)(width/2)  - (Max.x + Min.x) / 2.0 * Distance;
  Turtle.peny = (double)(height/2) - (Max.y + Min.y) / 2.0 * Distance;
  StkIdx = Turtle.dir = 0;
  Scaling = 0;

  /* do some changes */
  start_drawing();
  parse(&rules[1], &rules[rules[0]+1], depth);
  check_events(-1L);

  return 0;
}

void help(const char *name)
{
  fprintf(stderr, "\nUsage: %s [<options>] < <rules>\n\n", name);
  fprintf(stderr, "Options:\n");
  fprintf(stderr, "  -w <window width>\n");
  fprintf(stderr, "  -h <window height>\n\n"); 
  fprintf(stderr, "  -d <recursion depth> (0-%d)\n", MAX_DEPTH/2);
  fprintf(stderr, "  -a <angles> (360 / turn angle, 3-%d)\n", MAX_ANGLES);
  fprintf(stderr, "For rules (file) format see the manual page.\n\n");
  fprintf(stderr, "w/ Feb 1996 by Eero Tamminen\n");
}

/* Parse command line arguments. 1 for success, 0 for failure */
const char *args(int argc, char *argv[],
	int *depth, int *angle, int *width, int *height)
{
  int idx = 0;
  while(++idx < argc)
  {
    /* an one letter option and argument for it? */
    if(argv[idx][0] == '-' && argv[idx][2] == '\0' && idx+1 < argc)
      switch(argv[idx++][1])
      {
	case 'w':				/* window width */
	  if((*width = atoi(argv[idx])) < MIN_SIZE)
	    *width = MIN_SIZE;
	  break;
	case 'h':				/* window height */
	  if((*height = atoi(argv[idx])) < MIN_SIZE)
	    *height = MIN_SIZE;
	  break;
	case 'a':				/* 360 / turn angle */
          if((*angle = atoi(argv[idx])) < 3 || *angle > MAX_ANGLES)
	    return "invalid angles value";
	  break;
	case 'd':				/* recursion depth */
          if((*depth = atoi(argv[idx])) < 0)
	    return "bogus recursion depth";
	  break;
	default:
	  return "unrecognized option";
      }
    else
      return "not an option & it's argument";
  }
  return 0;	/* success */
}

/* parse rules (rule => 'char rule_lenght+2' 'char rule[]' with the
 * '=' character and white space stripped off from the input rule)
 * from the stdin to buffer
 */
char *get_rules(int *angle)
{
#define BUFFER_SIZE	512		/* the rules buffer size */
  static char buffer[BUFFER_SIZE];
  int idx = 1;

  scanf("%d", angle);

  if(*angle < 3 || *angle > MAX_ANGLES || scanf("%s", &buffer[idx]) != 1)
  {
    *angle = 6;
    /* no axiom -> default axiom & rule */
    memcpy(buffer, "\011" "f++f++f\0" "\013" "f" "f-f++f-f\0\0\0", 22);
    return buffer;
  }
  idx += buffer[idx-1] = strlen(&buffer[idx]) + 2;

  /* get the rules */
  while(idx+72 < BUFFER_SIZE)
  {
    if(scanf("%1s = %70s", &buffer[idx], &buffer[idx+1]) != 2)
      break;
    idx += buffer[idx-1] = strlen(&buffer[idx]) + 2;
  }
  buffer[idx] = 0;
  return buffer;
#undef BUFFER_SIZE
}


/* ---------------------- L-system parser ---------------------- */

/* axiom (intial rule), recursive rule, recursion depth */
void parse(const char *axiom, const char *rules, int depth)
{
  const char *rule;

  if(depth < 1)
  {
    while(*axiom)
      draw(*(axiom++));
    return;
  }

  while(*axiom)
  {
    rule = rules;
    while(*rule)
    {
      if(*axiom == *rule)
      {
        parse(&rule[1], rules, depth - 1);
	goto next;
      }
      rule += *(rule-1);
    }
    draw(*axiom);
    next:
    axiom++;
  }
}


/* --------------- Grammar -> graphics mapping -------------- */

void draw(char token)
{
  double oldx, oldy;

  switch(token)
  {
    case 'f':
    case 'F':
      oldx = Turtle.penx;
      oldy = Turtle.peny;
      Turtle.penx += Qcos[Turtle.dir] * Distance;
      Turtle.peny -= Qsin[Turtle.dir] * Distance;
      if(Scaling)
      {
       	if(Turtle.penx > Max.x) Max.x = Turtle.penx;
	if(Turtle.peny > Max.y) Max.y = Turtle.peny;
	if(Turtle.penx < Min.x) Min.x = Turtle.penx;
	if(Turtle.peny < Min.y) Min.y = Turtle.peny;
      }
      else if(token == 'f')
        w_line(Win, (short)oldx, (short)oldy,
	  (short)Turtle.penx, (short)Turtle.peny);
      break;

    case '+':			/* turn clockwise 1 'angle' */
      Turtle.dir++;
      if(Turtle.dir >= Angles)
        Turtle.dir = 0;
      break;

    case '-':			/* turn anti-clockwise 1 'angle' */
      Turtle.dir--;
      if(Turtle.dir < 0)
        Turtle.dir = Angles-1;
      break;

    case '[':			/* store turtle state to stack */
      if(StkIdx < MAX_DEPTH)
      {
	Store[StkIdx].penx = Turtle.penx;
	Store[StkIdx].peny = Turtle.peny;
	Store[StkIdx].dir  = Turtle.dir;
	StkIdx++;
      }
      break;

    case ']':			/* restore turtle state from stack */
      if(StkIdx > 0)
      {
	StkIdx--;
	Turtle.penx = Store[StkIdx].penx;
	Turtle.peny = Store[StkIdx].peny;
	Turtle.dir  = Store[StkIdx].dir;
      }
      break;
  }
}


/* ----------------------- W stuff --------------------------- */

/* initialize W connection and a output window */
const char *win_init(int width, int height)
{
  if(!w_init())
    return "can't connect to wserver";

  if(!(Win = w_create(width, height, WIN_PROPERTIES)))
    return "can't create window";

  /* draw white onto black window (looks IMHO better :-)) */
  w_pbox(Win, 0, 0, width, height);
  w_settitle(Win, " Scaling... ");

  if(w_open(Win, UNDEF, UNDEF) < 0)
    return "can't open window";

  return 0;
}

/* show that drawing has started... */
void start_drawing(void)
{
  /* change window title */
  w_settitle(Win, WIN_NAME);
  /* use white on black */
  w_setmode(Win, M_CLEAR);
}

/* wait until an event... */
void check_events(long timeout)
{
  WEVENT *ev;
  if(!(ev = w_queryevent(NULL, NULL, NULL, timeout)))
    return;

  if(ev->type == EVENT_GADGET &&
     (ev->key == GADGET_EXIT || ev->key == GADGET_CLOSE))
  {
    w_delete(Win);
    exit(0);
  }
}
